File: wrapper/java/src/main/java/io/github/snower/timslite/JournalQueueConsumer.java
package io.github.snower.timslite;

import io.github.snower.timslite.errors.TmslException;
import io.github.snower.timslite.uniffi.JournalQueueConsumerBridge;

/**
 * A consumer that reads journal change-log records within a specific consumer group.
 *
 * <p>{@code poll()} blocks until a record is available or the timeout expires.
 * After processing a record, call {@link #ack(long)} to advance the consumer
 * position. Unacknowledged records may be redelivered on the next poll.</p>
 *
 * <p>This class implements {@link AutoCloseable}.</p>
 */
public final class JournalQueueConsumer implements AutoCloseable {
    private static final java.lang.reflect.Method POLL_METHOD;

    static {
        try {
            POLL_METHOD = JournalQueueConsumerBridge.class.getMethod("poll-VKZWuLQ", long.class);
        } catch (NoSuchMethodException e) {
            throw new RuntimeException("Failed to find JournalQueueConsumerBridge.poll method", e);
        }
    }

    private final JournalQueueConsumerBridge bridge;
    private boolean closed;

    JournalQueueConsumer(JournalQueueConsumerBridge bridge) {
        this.bridge = bridge;
        this.closed = false;
    }

    /** Closes this journal consumer handle. */
    @Override
    public void close() {
        if (!closed) {
            closed = true;
            bridge.close();
        }
    }

    /**
     * Returns whether this journal consumer has been closed.
     *
     * @return {@code true} if {@link #close()} has been called
     */
    public boolean isClosed() {
        return closed;
    }

    private void checkNotClosed() {
        if (closed) {
            throw new IllegalStateException("JournalQueueConsumer is closed");
        }
    }

    /**
     * Polls for the next journal record, blocking up to the specified timeout.
     *
     * @param timeoutMs maximum time to wait in milliseconds; 0 returns
     *                  immediately if no record is available
     * @return the next journal record, or {@code null} if the timeout expires
     * @throws TmslException if polling fails
     */
    public JournalRecord poll(long timeoutMs) {
        checkNotClosed();
        try {
            io.github.snower.timslite.uniffi.JournalRecord kotlinRecord =
                    (io.github.snower.timslite.uniffi.JournalRecord) POLL_METHOD.invoke(bridge, timeoutMs);
            return kotlinRecord != null ? new JournalRecord(kotlinRecord) : null;
        } catch (java.lang.reflect.InvocationTargetException e) {
            Throwable cause = e.getCause();
            if (cause instanceof io.github.snower.timslite.uniffi.TmslException) {
                throw TmslException.fromUniFFI((io.github.snower.timslite.uniffi.TmslException) cause);
            }
            throw new RuntimeException(cause);
        } catch (IllegalAccessException e) {
            throw new RuntimeException("Failed to invoke poll", e);
        }
    }

    /**
     * Acknowledges a journal record, advancing the consumer position past its sequence.
     *
     * @param sequence the journal sequence number to acknowledge
     * @throws TmslException if the acknowledgment fails
     */
    public void ack(long sequence) {
        checkNotClosed();
        try {
            bridge.ack(sequence);
        } catch (io.github.snower.timslite.uniffi.TmslException e) {
            throw TmslException.fromUniFFI(e);
        }
    }
}

File: wrapper/dotnet/src/Timslite/JournalQueueConsumer.cs
using System;
using System.Threading;
using System.Threading.Tasks;
using Timslite.Errors;

namespace Timslite;

/// <summary>
/// A journal queue consumer that can poll for journal records and acknowledge processing.
/// </summary>
public sealed class JournalQueueConsumer : IDisposable
{
    private uniffi.timslite.JournalQueueConsumerBridge? _bridge;
    private bool _disposed;

    internal JournalQueueConsumer(uniffi.timslite.JournalQueueConsumerBridge bridge)
    {
        _bridge = bridge;
    }

    public void Dispose()
    {
        if (!_disposed)
        {
            _disposed = true;
            try { _bridge?.Release(); } catch { }
            _bridge?.Destroy();
            _bridge = null;
        }
        GC.SuppressFinalize(this);
    }

    public bool IsClosed => _disposed || (_bridge is null);

    private void CheckNotClosed()
    {
        if (_disposed || _bridge is null)
        {
            throw new ObjectDisposedException(nameof(JournalQueueConsumer), "Journal queue consumer is closed");
        }
    }

    /// <summary>
    /// Poll for the next journal record, blocking up to the specified timeout.
    /// Returns null if no record is available within the timeout.
    /// </summary>
    public JournalRecord? Poll(TimeSpan timeout)
    {
        CheckNotClosed();
        try
        {
            var result = _bridge!.Poll((ulong)(long)timeout.TotalMilliseconds);
            return result is null ? null : ConvertJournalRecord(result);
        }
        catch (uniffi.timslite.TmslException e)
        {
            throw TmslException.FromUniFFI(e);
        }
    }

    /// <summary>
    /// Asynchronously poll for the next journal record, blocking up to the specified timeout.
    /// Returns null if no record is available within the timeout.
    /// </summary>
    public Task<JournalRecord?> PollAsync(TimeSpan timeout, CancellationToken cancellationToken = default)
    {
        return Task.Run(() =>
        {
            cancellationToken.ThrowIfCancellationRequested();
            return Poll(timeout);
        }, cancellationToken);
    }

    /// <summary>
    /// Acknowledge that a journal record has been processed.
    /// </summary>
    public void Ack(long sequence)
    {
        CheckNotClosed();
        try
        {
            _bridge!.Ack(sequence);
        }
        catch (uniffi.timslite.TmslException e)
        {
            throw TmslException.FromUniFFI(e);
        }
    }

    private static JournalRecord ConvertJournalRecord(uniffi.timslite.JournalRecord r)
    {
        return new JournalRecord(r.Sequence, r.Data);
    }
}
